controllers/k8ssandra/k8ssandracluster_controller_test.go

```diff
@@ -970,6 +971,154 @@ func setReplicationStatusDone(ctx context.Context, t *testing.T, f *framework.Fr
+	verifySystemReplicationAnnotationSet(ctx, t, f, kc)
+
+	t.Log("check that dc1 was created")
+	require.Eventually(f.DatacenterExists(ctx, dc1Key), timeout, interval)
+
+	t.Log("update datacenter status to scaling up")
+	err = f.PatchDatacenterStatus(ctx, dc1Key, func(dc *cassdcapi.CassandraDatacenter) {
+		dc.SetCondition(cassdcapi.DatacenterCondition{
+			Type:               cassdcapi.DatacenterScalingUp,
+			Status:             corev1.ConditionTrue,
+			LastTransitionTime: metav1.Now(),
+		})
+	})
+	require.NoError(err, "failed to patch datacenter status")
+
+	t.Log("check that the K8ssandraCluster status is updated")
+	require.Eventually(func() bool {
+		kc := &api.K8ssandraCluster{}
+		err = f.Get(ctx, kcKey, kc)
+		if err != nil {
+			t.Logf("failed to get K8ssandraCluster: %v", err)
+			return false
+		}
+
+		if (&kc.Status).GetConditionStatus(api.CassandraInitialized) == corev1.ConditionTrue {
+			t.Logf("Did not expect status condition %s to be true", api.CassandraInitialized)
+			return false
+		}
+
+		if len(kc.Status.Datacenters) == 0 {
+			return false
+		}
+
+		k8ssandraStatus, found := kc.Status.Datacenters[dc1Key.Name]
+		if !found {
+			t.Logf("status for datacenter %s not found", dc1Key)
+			return false
+		}
+
+		condition := FindDatacenterCondition(k8ssandraStatus.Cassandra, cassdcapi.DatacenterScalingUp)
+		return !(condition == nil && condition.Status == corev1.ConditionFalse)
+	}, timeout, interval, "timed out waiting for K8ssandraCluster status update")
+
+	t.Log("update dc1 status to ready")
+	err = f.SetDatacenterStatusReady(ctx, dc1Key)
+	require.NoError(err, "failed to update dc1 status to ready")
+
+	t.Log("check that stargate sg1 is created")
+	require.Eventually(f.StargateExists(ctx, sg1Key), timeout, interval)
+
+	t.Log("check that cass DC sets allow_alter_rf_during_range_movement")
+	cassDc := &cassdcapi.CassandraDatacenter{}
+	err = f.Get(ctx, dc1Key, cassDc)
+	require.NoError(err, "failed to get CassandraDatacenter")
+	dcConfig, err := utils.UnmarshalToMap(cassDc.Spec.Config)
+	require.NoError(err, "failed to unmarshall CassandraDatacenter config")
+	jvmOpts := dcConfig["cassandra-env-sh"].(map[string]interface{})["additional-jvm-opts"].([]interface{})
+	require.Contains(jvmOpts, "-Dcassandra.allow_alter_rf_during_range_movement=true")
+
+	err = f.Get(ctx, kcKey, kc)
+	require.NoError(err, "failed to get K8ssandraCluster")
+	dcGeneration := kc.Status.Datacenters["dc1"].Cassandra.ObservedGeneration
+
+	t.Log("remove stargate sg1 from kc spec")
+	patch := client.MergeFromWithOptions(kc.DeepCopy(), client.MergeFromWithOptimisticLock{})
+	kc.Spec.Cassandra.Datacenters[0].Stargate = nil
+	err = f.Client.Patch(ctx, kc, patch)
+	require.NoError(err, "failed to update K8ssandraCluster")
+
+	t.Log("check that stargate sg1 is deleted")
+	require.Eventually(func() bool {
+		err = f.Get(ctx, sg1Key, &stargateapi.Stargate{})
+		return errors.IsNotFound(err)
+	}, timeout, interval)
```

Comment:
Is this enough to guarantee that I will observe the latest versions of the `K8ssandraCluster` and `cassdcapi.CassandraDatacenter` in the lines that follow, or should I wait on another condition?

Reply:
The check is fine. Keep in mind that the controllers using a caching client by default. They perform reads from a cache which means that they can be stale. Tests however are configure client a client that reads directly from the API server which means they they will always observe the latest version with respect to what is in etcd. Polling as you have done here with the `Eventually` call is needed when we need to wait for changes to be made by the controller, i.e., desired state --> actual state.